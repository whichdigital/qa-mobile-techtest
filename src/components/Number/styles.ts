import { StyleSheet } from 'react-native';

export default StyleSheet.create({
    container: {
        flex: 1,
        alignItems: 'center',
        justifyContent: 'center',
        height: 100,
    },
    number: {
        width: '80%',
        height: '80%',
        alignItems: 'center',
        justifyContent: 'center',
    },
    value: {
        borderColor: 'red',
        borderWidth: 2,
    },
    active: {
        backgroundColor: '#7ED6BE',
    },
    multiple: {
        backgroundColor: '#FEEBA7',
        borderRadius: 5,
    },
});
