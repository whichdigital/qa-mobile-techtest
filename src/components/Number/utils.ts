import { AccessibilityState } from 'react-native';

export function isMultipleOf(a: number, b: null | number): boolean {
    if (!b) {
        return false;
    }

    return a % b === 0;
}

export function getAccessibilityState(a: number, b: null | number): AccessibilityState {
    if (a === b) {
        return { selected: true, checked: true };
    }

    if (isMultipleOf(a, b)) {
        return { checked: true };
    }

    return {};
}
