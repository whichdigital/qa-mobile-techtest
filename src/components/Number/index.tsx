import React from 'react';
import { ReactElement } from 'react';
import { Text, View, TouchableOpacity } from 'react-native';

import { Selected, SetSelected } from '../../types';
import styles from './styles';
import { getAccessibilityState, isMultipleOf } from './utils';

type Props = {
    value: number;
    selected: Selected;
    onPress: SetSelected;
};

export default function Number({ value, selected, onPress }: Props): ReactElement {
    return (
        <TouchableOpacity
            testID="pressable"
            style={styles.container}
            onPress={() => onPress(value === selected ? null : value)}>
            <View
                testID="number"
                style={[
                    styles.number,
                    isMultipleOf(value, selected) && styles.multiple,
                    value === selected && styles.active,
                ]}>
                <Text accessibilityState={getAccessibilityState(value, selected)}>{value}</Text>
            </View>
        </TouchableOpacity>
    );
}
