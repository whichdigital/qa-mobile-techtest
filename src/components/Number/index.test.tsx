import { fireEvent, render } from '@testing-library/react-native';
import React from 'react';

import Number from '.';

describe('<Number />', () => {
    it('should be able to invoke the `onPress` handler when clicking on a number', () => {
        const spies = { onPress: jest.fn() };

        const { getByTestId } = render(<Number value={5} selected={null} onPress={spies.onPress} />);

        fireEvent.press(getByTestId('pressable'));
        expect(spies.onPress).toHaveBeenCalledTimes(1);
    });

    it('should be able to supply the appropriate styles', () => {
        const { rerender, getByTestId } = render(<Number value={5} selected={null} onPress={() => {}} />);

        expect(getByTestId('number').props.style).toMatchSnapshot();

        rerender(<Number value={20} selected={5} onPress={() => {}} />);
        expect(getByTestId('number').props.style).toMatchSnapshot();

        rerender(<Number value={5} selected={5} onPress={() => {}} />);
        expect(getByTestId('number').props.style).toMatchSnapshot();
    });
});
