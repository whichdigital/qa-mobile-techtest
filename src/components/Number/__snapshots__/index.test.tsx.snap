// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Number /> should be able to supply the appropriate styles 1`] = `
Array [
  Object {
    "alignItems": "center",
    "height": "80%",
    "justifyContent": "center",
    "width": "80%",
  },
  false,
  false,
]
`;

exports[`<Number /> should be able to supply the appropriate styles 2`] = `
Array [
  Object {
    "alignItems": "center",
    "height": "80%",
    "justifyContent": "center",
    "width": "80%",
  },
  Object {
    "backgroundColor": "#FEEBA7",
    "borderRadius": 5,
  },
  false,
]
`;

exports[`<Number /> should be able to supply the appropriate styles 3`] = `
Array [
  Object {
    "alignItems": "center",
    "height": "80%",
    "justifyContent": "center",
    "width": "80%",
  },
  Object {
    "backgroundColor": "#FEEBA7",
    "borderRadius": 5,
  },
  Object {
    "backgroundColor": "#7ED6BE",
  },
]
`;
